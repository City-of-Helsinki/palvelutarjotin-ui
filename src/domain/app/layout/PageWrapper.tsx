import { useTranslation } from 'next-i18next';
import { useRouter } from 'next/router';
import React from 'react';

import PageMeta, {
  getLanguageAwarePath,
} from '../../../common/components/meta/PageMeta';
import { SUPPORTED_LANGUAGES } from '../../../constants';
import { Language } from '../../../types';

interface Props {
  title?: string;
}

const PageWrapper: React.FC<Props> = ({ children, title = 'appName' }) => {
  const { t } = useTranslation();
  const { asPath } = useRouter();

  const translatedTitle =
    title !== 'appName'
      ? `${t(title)} - ${t('common:appName')}`
      : t('common:appName');

  return (
    <div>
      <PageMeta
        title={translatedTitle}
        localePaths={Object.values(SUPPORTED_LANGUAGES).map((lang) => ({
          locale: lang as Language,
          path: getLanguageAwarePath(lang, asPath),
        }))}
      />
      <main>{children}</main>
    </div>
  );
};

export default PageWrapper;
