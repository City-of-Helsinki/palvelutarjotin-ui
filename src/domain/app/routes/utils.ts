import { CollectionItemType } from 'react-helsinki-headless-cms';

import { ROUTES } from './constants';

export const getCmsPagePath = (uri?: string | null): string => {
  return ROUTES.CMS_PAGE.replace('/:uri', uri ?? '');
};

export const getCmsArticlePath = (uri?: string | null): string => {
  return ROUTES.CMS_ARTICLE.replace('/:uri', uri ?? '');
};

export const getCollectionItemUrl = (item: CollectionItemType): string => {
  if (!item) {
    return '#';
  }
  if (item.__typename === 'Post') {
    return getCmsArticlePath(item.uri);
  }
  if (item.__typename === 'Page') {
    return getCmsPagePath(item.uri);
  }
  return item?.uri ?? '';
};
