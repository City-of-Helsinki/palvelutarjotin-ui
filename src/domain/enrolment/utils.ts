import {
  EnrolOccurrenceMutationInput,
  Language,
  NotificationType,
} from '../../generated/graphql';
import { EnrolmentFormFields } from './enrolmentForm/constants';

const getNotificationType = (values: EnrolmentFormFields): NotificationType => {
  return values.hasEmailNotification && values.hasSmsNotification
    ? NotificationType.EmailSms
    : NotificationType.Email;
};

export const getEnrolmentPayload = ({
  occurrenceIds,
  values,
}: {
  occurrenceIds: string[];
  values: EnrolmentFormFields;
}): EnrolOccurrenceMutationInput => {
  return {
    occurrenceIds,
    notificationType: getNotificationType(values),
    person: values.isSameResponsiblePerson
      ? undefined
      : { ...values.person, language: values.language as Language },
    studyGroup: {
      ...values.studyGroup,
      amountOfAdult: Number(values.studyGroup.amountOfAdult),
      groupSize: Number(values.studyGroup.groupSize),
      person: {
        ...values.studyGroup.person,
        language: values.language as Language,
      },
      studyLevels: values.studyGroup.studyLevels,
    },
  };
};

declare let grecaptcha: ReCAPTCHA;

interface ReCAPTCHA {
  ready: (f: () => void) => void;
  execute: (key: string, action: { action: string }) => Promise<string>;
}

export const getCAPTCHAToken = (): Promise<string> => {
  return new Promise((resolve, reject) => {
    if (process.env.NEXT_PUBLIC_CAPTCHA_KEY) {
      const captchaKey = process.env.NEXT_PUBLIC_CAPTCHA_KEY;
      grecaptcha.ready(() => {
        grecaptcha
          .execute(captchaKey, { action: 'submit' })
          .then(resolve)
          .catch(reject);
      });
    } else {
      reject('CAPTCHA_KEY missing!');
    }
  });
};
