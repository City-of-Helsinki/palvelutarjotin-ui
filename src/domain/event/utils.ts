import { isToday, isTomorrow } from 'date-fns';
import { TFunction } from 'next-i18next';

import { EventFieldsFragment } from '../../generated/graphql';
import { Language } from '../../types';
import formatDate from '../../utils/formatDate';
import getLocalisedString from '../../utils/getLocalisedString';
import getTimeFormat from '../../utils/getTimeFormat';
import { EVENT_PLACEHOLDER_IMAGES } from './constants';

/**
 * Get event placeholder image url
 * @param {string} id
 * @return {string}
 */
export const getEventPlaceholderImage = (id: string): string => {
  const numbers = id.match(/\d+/g);
  const sum = numbers
    ? numbers.reduce((prev: number, cur: string) => prev + Number(cur), 0)
    : 0;
  const index = sum % 4;

  return EVENT_PLACEHOLDER_IMAGES[index];
};

export const getEventStartTimeStr = (
  event: EventFieldsFragment,
  locale: Language,
  t: TFunction
): string | null => {
  const startTime = event.startTime ? new Date(event.startTime) : null;
  const timeFormat = getTimeFormat(locale);

  if (!startTime) return null;

  if (isToday(startTime))
    return t('event:eventCard.startTime.today', {
      time: formatDate(startTime, timeFormat, locale),
    });

  if (isTomorrow(startTime))
    return t('event:eventCard.startTime.tomorrow', {
      time: formatDate(startTime, timeFormat, locale),
    });

  return t('event:eventCard.startTime.other', {
    date: formatDate(startTime),
    time: formatDate(startTime, timeFormat, locale),
  });
};

/**
 * Check is event free
 * @param eventData
 * @return {boolean}
 */
export const isEventFree = (event: EventFieldsFragment): boolean =>
  Boolean(event.offers.find((item) => item.isFree)?.isFree);

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const getEventFields = (
  event: EventFieldsFragment | undefined | null,
  locale: Language
) =>
  event
    ? {
        isEventFree: isEventFree(event),
        eventName: getLocalisedString(event?.name, locale),
        shortDescription: getLocalisedString(event?.shortDescription, locale),
        description: getLocalisedString(event?.description, locale),
        imageUrl: event?.images?.[0]?.url,
        imageAltText: event?.images?.[0]?.altText,
        photographerName: event?.images?.[0]?.photographerName,
        organization: event.pEvent?.organisation?.name,
        contactPhoneNumber: event.pEvent?.contactPhoneNumber,
        contactEmail: event.pEvent?.contactEmail,
        contactPerson: event.pEvent?.contactPerson?.name,
      }
    : {};
