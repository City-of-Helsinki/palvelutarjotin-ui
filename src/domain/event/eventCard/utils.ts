import { TFunction } from 'next-i18next';

import { EnrolmentStatusKeywordPropsType } from '../../../common/components/keyword/EnrolmentStatusKeyword';
import { EventsFieldsFragment } from '../../../generated/graphql';
import { Language } from '../../../types';
import { formatDateRange } from '../../../utils/time/format';
import { isMultidayOccurrence } from '../../occurrence/utils';
import { getEventStartTimeStr } from '../utils';

export const getOccurrenceDateStr = (
  occurrence: { startTime: string; endTime: string },
  locale: Language,
  t: TFunction
): string | null => {
  if (isMultidayOccurrence(occurrence)) {
    return formatDateRange(
      new Date(occurrence.startTime),
      new Date(occurrence.endTime)
    );
  }

  return getEventStartTimeStr(new Date(occurrence.startTime), locale, t);
};

/**
 * Has the event enrolments enabled in any of it's occurrences?
 * @param event The event to check the status for.
 * @returns `true` if any of the event occurrences has enrolments and
 *  `false` if none.
 */
export const hasEventEnrolmentsEnabled = (event: EventsFieldsFragment) =>
  Boolean(event.pEvent.enrolmentStart) || hasEventExternalEnrolments(event);

/**
 * Has the event external enrolments enabled in any of it's occurrences?
 *
 * @param event The event to check the status for.
 * @returns `true` if the event has some external enrolments and
 *  `false` if none of the occurrences has external enrolments.
 */
export const hasEventExternalEnrolments = (event: EventsFieldsFragment) =>
  Boolean(event.pEvent.externalEnrolmentUrl);

/**
 * Resolve the enrolment status of an event. Has the event free spots available?
 *
 * @param event The event to check the status for.
 * @returns Returns `null` if the enrolment status could not be resolved or
 *    the event has no enrolments at all,
 *    `true` if there are any free spots available and `false` if there are no
 *    spots available.
 */
export const hasEventFreeSpotsAvailable = (
  event: EventsFieldsFragment
): boolean | null => event.pEvent.hasSpaceForEnrolments ?? null;

/**
 * Determines the enrollment status of an event.
 *
 * @param event The event to check the enrollment status for.
 * @returns The enrollment status of the event, or `undefined`
 *      if no enrolments exist or the status could not be determined.
 */
export const getEventEnrolmentStatus = (
  event: EventsFieldsFragment
): EnrolmentStatusKeywordPropsType['enrolmentStatus'] => {
  if (hasEventEnrolmentsEnabled(event)) {
    if (hasEventExternalEnrolments(event)) {
      return 'externally_unknown';
    }
    const hasEventFreeSpots = hasEventFreeSpotsAvailable(event);

    if (hasEventFreeSpots === false) {
      // Event has no free spots available, so it is full.
      return 'full';
    } else if (hasEventFreeSpots === true) {
      // Event has some free spots available.
      return 'free_space_available';
    }
  }
  // No enrolments at all or the enrolment status could not be resolved
  return undefined;
};
