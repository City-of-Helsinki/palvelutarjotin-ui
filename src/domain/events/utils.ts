import isValidDate from 'date-fns/isValid';

import { EVENT_LANGUAGES } from '../../constants';
import {
  EventsQueryVariables,
  EventFieldsFragment,
  EventsQuery,
} from '../../generated/graphql';
import { queryParameterToArray } from '../../utils/queryParameterToArray';
import { EventSearchFormValues } from './eventSearchForm/EventSearchForm';

export const getTextFromDict = (
  query: NodeJS.Dict<string | string[]>,
  key: string,
  defaultValue = ''
): string | undefined => {
  const value = query[key];
  if (Array.isArray(value)) return value.join(',');
  return (value as string) || defaultValue;
};

type EventFilterOptions = {
  page?: number;
  pageSize?: number;
};

export const getEventFilterVariables = (
  query: NodeJS.Dict<string | string[]>,
  options?: EventFilterOptions
): EventsQueryVariables => ({
  include: ['keywords,location'],
  text: getTextFromDict(query, 'text', undefined),
  inLanguage: getTextFromDict(query, 'inLanguage', undefined),
  start:
    typeof query.date === 'string' && isValidDate(new Date(query.date))
      ? new Date(query.date).toISOString()
      : null,
  ...options,
});

export const getInitialValues = (
  query: NodeJS.Dict<string | string[]>
): EventSearchFormValues => {
  return {
    text: getTextFromDict(query, 'text') || '',
    date: getInitialDate(query.date),
    inLanguage: queryParameterToArray(query.inLanguage as EVENT_LANGUAGES),
  };
};

export const getInitialDate = (date?: string | string[]): Date | null => {
  if (typeof date === 'string' && isValidDate(new Date(date))) {
    return new Date(date);
  }

  return null;
};

export const getEventsThatHaveUpcomingOccurrence = (
  eventData?: EventsQuery
): EventFieldsFragment[] | undefined => {
  return eventData?.events?.data.filter((e) => e.pEvent.nextOccurrenceDatetime);
};
