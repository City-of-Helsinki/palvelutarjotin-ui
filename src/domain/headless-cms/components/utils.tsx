import { randomUUID } from 'crypto';

import React from 'react';
import {
  ArticleType,
  Card,
  Collection,
  GeneralCollectionType,
  getCollectionCards,
  PageType,
} from 'react-helsinki-headless-cms';

import { getCmsArticlePath, getCmsPagePath } from '../../app/routes/utils';

export function getCmsCollectionList(
  collections: GeneralCollectionType[]
): React.ReactElement<typeof Collection>[] {
  return collections.map((collection) => (
    <Collection
      key={`collection-${randomUUID()}`}
      title={collection.title}
      collectionContainerProps={{
        withDots: collection.items.length < 4 ? false : true,
      }}
      type="grid"
      cards={getCollectionCards(collection as GeneralCollectionType, []).map(
        (item) => (
          <Card
            key={`card-${item.id}`}
            {...item}
            clampText={true}
            withShadow={true}
            hasLink={true}
            url={getCollectionItemUrl(item as PageType)}
            // imageLabel={item.featuredImage?.node?.title ?? ''}
            // imageUrl={item.featuredImage?.node?.mediaItemUrl ?? ''}
          />
        )
      )}
    />
  ));
}

export function getCollectionItemUrl(item?: PageType | ArticleType): string {
  if (!item) {
    return '#';
  }
  if (item.__typename === 'Post') {
    return getCmsArticlePath(item.uri);
  }
  if (item.__typename === 'Page') {
    return getCmsPagePath(item.uri);
  }
  return item?.uri ?? '';
}
