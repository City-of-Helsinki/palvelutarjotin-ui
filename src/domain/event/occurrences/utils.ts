import {
  EventFieldsFragment,
  OccurrenceFieldsFragment,
} from '../../../generated/graphql';
import { ENROLMENT_ERRORS } from '../../enrolment/constants';
import {
  hasOccurrenceSpace,
  isEnrolmentClosed,
  isOccurrenceCancelled,
} from '../../occurrence/utils';

export function getEnrolmentError(
  occurrence: OccurrenceFieldsFragment,
  event: EventFieldsFragment
): ENROLMENT_ERRORS | null {
  if (isEnrolmentClosed(occurrence, event))
    return ENROLMENT_ERRORS.ENROLMENT_CLOSED_ERROR;
  if (!hasOccurrenceSpace(occurrence))
    return ENROLMENT_ERRORS.NOT_ENOUGH_CAPACITY_ERROR;
  if (isOccurrenceCancelled(occurrence)) {
    return ENROLMENT_ERRORS.ENROLMENT_CANCDELLED_ERROR;
  }

  return null;
}

export function getOrderedLanguages(
  languages: { id: string; name: string }[]
): string[] {
  const languageOrder = ['fi', 'sv', 'en'];
  const languageStrings = languages.map((l) => l.id.split('_')[0]);

  // order languages based on languageOrder, other languages go
  // to the end of the array
  return languageStrings.sort((a, b) => {
    const aIndex = languageOrder.indexOf(a);
    const bIndex = languageOrder.indexOf(b);

    if (aIndex > -1 && bIndex > -1) {
      return aIndex - bIndex;
    }

    if (aIndex > -1 && bIndex === -1) {
      return -1;
    }

    if (bIndex > -1 && aIndex === -1) {
      return 1;
    }

    return 0;
  });
}
