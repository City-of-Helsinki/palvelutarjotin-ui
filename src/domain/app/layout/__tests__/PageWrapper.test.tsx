/* eslint-disable @typescript-eslint/no-explicit-any */
import matomo from '@datapunt/matomo-tracker-react';
import * as nextRouter from 'next/router';
import * as React from 'react';

import { render } from '../../../../utils/testUtils';
import PageWrapper from '../PageWrapper';

jest.mock('@datapunt/matomo-tracker-react', () => ({
  useMatomo: jest.fn().mockReturnValue({ trackPageView: jest.fn() }),
}));

const setHref = (href: string) => {
  Object.defineProperty(window, 'location', {
    value: {
      ...window.location,
      href,
    },
    writable: true,
  });
};

test('PageWrapper matches snapshot', async () => {
  const { container } = render(
    <PageWrapper>
      <div>Children</div>
    </PageWrapper>
  );

  expect(container).toMatchSnapshot();
});

test('trackPageView gets called when pathname changes', () => {
  const testHref1 = 'testurl.com';
  const testHref2 = 'testurl2.com';
  const trackPageViewMock = jest.fn();
  setHref(testHref1);

  jest
    .spyOn(matomo, 'useMatomo')
    .mockReturnValue({ trackPageView: trackPageViewMock } as any);

  const { rerender } = render(<PageWrapper />, { path: '/test1' });

  expect(trackPageViewMock).toHaveBeenCalledWith({ href: testHref1 });

  trackPageViewMock.mockReset();

  setHref(testHref2);
  jest
    .spyOn(nextRouter, 'useRouter')
    .mockReturnValue({ asPath: '/test2' } as any);

  rerender(<PageWrapper />);

  expect(trackPageViewMock).toHaveBeenCalledWith({ href: testHref2 });
});
