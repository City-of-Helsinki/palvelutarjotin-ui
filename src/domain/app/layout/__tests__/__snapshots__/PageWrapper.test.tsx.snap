// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageWrapper matches snapshot 1`] = `
<div>
  <div>
    <main>
      <div>
        Children
      </div>
    </main>
  </div>
</div>
`;
